services/payments/statemachine.go

```diff
@@ -218,20 +220,24 @@ func (s *Service) DriveTransaction(
 		// current persisted state
 		history, err := s.datastore.GetPaymentStateHistory(ctx, state.DocumentID)
 		if err != nil {
-			return fmt.Errorf("failed to get history from document id", err)
+			return fmt.Errorf("failed to get history from document id: %w", err)
 		}
 		persistedState, err := history.GetAuthenticatedPaymentState(
 			s.verifierStore,
 			state.DocumentID,
 		)
 		if err != nil {
-			return fmt.Errorf("failed to validate payment state history", err)
+			return fmt.Errorf("failed to validate payment state history: %w", err)
 		}
-		// If there is idempotency data in qldb and it is different from the idempotency data in the
-		// current state it means that there was a race between two calls to Authenticate and we are
-		// operating on the loser. There is no risk to proceeding as long as we retain the winner
-		// idempotency.
-		if persistedState != nil && !bytes.Equal(state.ExternalIdempotency, persistedState.ExternalIdempotency) {
+		// If there is idempotency data in qldb and in our generated state and
+		// it is different from the idempotency data in the current state it
+		// means that there was a race between two calls to Authenticate and we
+		// are operating on the loser. There is no risk to proceeding as long as
+		// we retain the winner idempotency.
+		if persistedState != nil &&
+			persistedState.ExternalIdempotency != nil &&
+			state.ExternalIdempotency != nil &&
+			!bytes.Equal(state.ExternalIdempotency, persistedState.ExternalIdempotency) {
```

Comment:
At some point (soon) it might be good to extract this into a separate predicate to be able to test it in isolation.